lib/array.js

```diff
@@ -1,47 +1,30 @@
 'use strict';
 
+const common = require('@metarhia/common');
+const { asyncIter } = require('./async-iterator.js');
+const { promisify } = require('util');
+
 // Asynchronous map (iterate parallel)
-//   items - <Array>, incoming
+//   items - <Iterable>, incoming
 //   fn - <Function>, to be executed for each value in the array
 //     current - <any>, current element being processed in the array
 //     callback - <Function>
 //       err - <Error> | <null>
 //       value - <any>
 //   done - <Function>, on done
 //     err - <Error> | <null>
-//     result - <Array>
-const map = (items, fn, done) => {
-  const len = items.length;
-  if (!len) {
-    done(null, []);
-    return;
-  }
-  let errored = false;
-  let count = 0;
-  const result = new Array(len);
-
-  const next = (index, err, value) => {
-    if (errored) return;
-    if (err) {
-      errored = true;
-      done(err);
-      return;
-    }
-    result[index] = value;
-    count++;
-    if (count === len) done(null, result);
-  };
-
-  for (let i = 0; i < len; i++) {
-    fn(items[i], next.bind(null, i));
-  }
+//     result - <Iterable>
+const map = (items, fn, done = common.emptiness) => {
+  const isArray = Array.isArray(items);
+  asyncIter(items)
```

Comment:
I think wrapping everything in `AsyncIterator` and additionally using `util.promisify()` adds unnecessary overhead, that can be avoided by just rewriting the original functions with iterators in mind. This also won't require any changes to `metasync.AsyncIterator`. If we decide to keep it the way it is in this PR we should probably at least benchmark this variant to make sure this doesn't lead to significant performance regressions.  
/cc @tshemsedinov